data_model/src/lib.rs

```diff
@@ -32,7 +33,78 @@ pub mod transaction;
 
 /// `Name` struct represents type for Iroha Entities names, like [`Domain`](`domain::Domain`)'s name or [`Account`](`account::Account`)'s
 /// name.
-pub type Name = String;
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Decode,
+    Encode,
+    Deserialize,
+    Serialize,
+    IntoSchema,
+)]
+pub struct Name(String);
+
+impl Name {
+    /// Construct [`Name`] if `name` is valid.
+    ///
+    /// # Errors
+    /// Fails if parsing fails
+    #[inline]
+    pub fn new(name: &str) -> Result<Self> {
+        name.parse::<Self>()
+    }
+
+    /// Instantly construct [`Name`] assuming `name` is valid.
+    #[inline]
+    #[allow(clippy::expect_used)]
+    pub fn test(name: &str) -> Self {
+        name.parse::<Self>()
+            .expect("Valid names never fail to parse")
+    }
+
+    /// Provide an access to the inner `String` of this [`Name`].
+    pub fn inner(&self) -> &String {
```

Comment:
this is perfect for `AsRef<str>`

Reply:
It requires `dyn` though. Isn't `&String` enough?

Reply:
```suggestion
impl AsRef<str> for Name {
    pub fn as_ref(&self) -> &str {
        self.0.as_str()
    }
}
```

this should work

Reply:
but `&String` would be ok. If it's somehow required on the call site?